### PR TITLE
Refactored webviewer page

### PR DESCRIPTION
**Pull request by micky2be**
Went through some "big" refactoring.

This mostly include the following changes:
- avoid too many DOM requests
  - use data attributes when needed instead of parsing IDs
- use of JavaScript data (created a mini store system)
- avoid re-creating the image table when not necessary
- default to a placeholder images if not created yet (404)
- avoid creating and relying on public functions
- some linting

@mcmonkey4eva I'm not quite sure if we want to use or remove jQuery (I'm open to both, DOM API is more verbose though). 
This refactoring was made to allow further update, and try to rely less on generated HTML from python.
_I also worked in parallel on a fix for gradio to avoid the nasty 500 errors for non existing images._

Let me know if you have any questions or concerns

### CODE REVIEW

File: gridgencore.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 ASSET_DIR = os.path.dirname(__file__) + "/assets"
 EXTRA_FOOTER = "..."
-EXTRA_ASSETS = []
+EXTRA_ASSETS = ["placeholder-landscape.png", "placeholder-portrait.png", "placeholder-square.png"]
```
Comment (mcmonkey4eva) [inline gridgencore.py:11]: This should not be here, it should be on line 583 (same place you added `styles.css`)

File: gridgencore.py [hunk 2]
```diff
@@ -510,7 +510,7 @@ def get_axis(axis: str):
         return json.dumps(result)
 
     def radio_button_html(name, id, descrip, label):
-        return f'<input type="radio" class="btn-check" name="{name}" id="{str(id).lower()}" autocomplete="off" checked=""><label class="btn btn-outline-primary" for="{str(id).lower()}" title="{descrip}">{label}</label>\n'
```
Comment (mcmonkey4eva) [inline gridgencore.py:513]: Why was this removed? iirc that was very intentionally added as a bug-fix.

File: gridgencore.py [hunk 3]
```diff
@@ -544,10 +542,10 @@ def build_html(grid):
...
-                advanced_settings += f'&nbsp;&nbsp;<button class="submit" onclick="javascript:toggleShowAllAxis(\'{axis.id}\')">Toggle All</button>'
-                content += '</ul>\n<div class="tab-content">\n'
+                    content += f'\t<button class="nav-link{active}" data-bs-toggle="pill" data-bs-target="#tab_{axis.id}__{val.key}" data-axis-id="{axis.id}" data-val-key="{val.key}" aria-selected="{selected}" role="tab" title="{val.title}: {descrip}">{val.title}</button>\n'
+                    advanced_settings += f'&nbsp;<input class="form-check-input" type="checkbox" autocomplete="off" id="showval_{axis.id}__{val.key}" data-axis-id="{axis.id}" data-val-key="{val.key}" checked="true"> <label class="form-check-label" for="showval_{axis.id}__{val.key}" title="Uncheck this to hide \'{val.title}\' from the page.">{val.title}</label>'
+                advanced_settings += f'&nbsp;&nbsp;<button class="submit" data-axis-id={axis.id}>Toggle All</button>'
```
Comment (mcmonkey4eva) [inline gridgencore.py:547]: this is missing quotes

File: assets/styles.css [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+}
+.nav-pills .nav-link.disabled, .nav-pills .nav-link:disabled {
+    color: var(--bs-nav-link-disabled-color);
+    background-color: transparent;
+    border-color: transparent;
+}
+.nav-pills .nav-link.active,
+.nav-pills .show > .nav-link {
+    border-color: var(--bs-nav-pills-link-active-border-color);
+}
```
Comment (mcmonkey4eva) [inline assets/styles.css:135]: NL@EOF

File: assets/page.html [hunk 5]
```diff
@@ -3,106 +3,9 @@
     <meta charset="utf-8" />
     <meta name="viewport" content="width=device-width, initial-scale=1.0" />
     <link rel="stylesheet" href="bootstrap.min.css" media="screen">
+    <link rel="stylesheet" href="styles.css" media="screen">
```
Comment (mcmonkey4eva) [inline assets/page.html:6]: If this is going to be moved to a separate file it needs a `?vary=#` on it (this is a way to workaround caching issues).

File: assets/page.html [hunk 6]
```diff
@@ -160,10 +77,9 @@ <h2 class="accordion-header" id="setting_accordion_heading">
             <br>{EXTRA_FOOTER}
             <br>Made using the <a href="https://bootswatch.com/darkly/">Darkly Bootstrap Theme</a> by Thomas Park, which was released under the <a href="https://github.com/thomaspark/bootswatch/blob/95df99d76147797cbcb1014b639805add2327f65/LICENSE">MIT License</a>.
             <br>
-            <br><div style="width:256px;height:512px;"></div>
```
Comment (mcmonkey4eva) [inline assets/page.html:163]: Why was this removed? This was a very intentional bugfix.

File: assets/a1111webui.js [hunk 7]
```diff
@@ -2,70 +2,117 @@
...
-function formatMetadata(valSet) {
-    var count = Object.keys(valSet).length;
-    if (count == 0) {
-        return "";
-    }
-    else if (count == 1) {
-        return valSet["error"];
+'use strict';
+
+(() => {
```
Comment (mcmonkey4eva) [inline assets/a1111webui.js:7]: ... What. Why. What.

File: assets/a1111webui.js [hunk 8]
```diff
@@ -2,70 +2,117 @@
...
+            formatMet('Variation seed', valSet['varseed'], '0'),
+            formatMet('Variation seed strength', valSet['varstrength'], '0'),
+            // Seed resize from
+            formatMet('Denoising strength', valSet['denoising']),
+            // Conditional mask weight
+            formatMet('Eta', valSet['eta']),
+            formatMet('Clip skip', valSet['clipskip'], '1'),
+            // ENSD
+        ].filter(Boolean);
+        metadataSet.push(keyData.join(', '));
```
Comment (mcmonkey4eva) [inline assets/a1111webui.js:59]: This... does the same thing with several extra steps? Why?

**Comment by mcmonkey4eva** [comment]
That's, uh, a lot.

Straightaway the main issue I have here is that you changed so much at once it's hard to track what anything is or why. It would have been much better to submit as separate PRs, or at least split into separate commits.

There are several changes that just reformat entire files for no reason, eg tabbing out the javascript and changing the quotes from `"` to `'` ... which are just git changes for the sake of git changes, with no identifiable real value.

I don't think I can comfortably pull this for so long as the git commit is unreadable from changing too much at once. I'd suggest you discard the tabulation/quote/ordering changes in the JS so it can at least be focused down to actual applied changes.

Also, I don't think 3x placeholder images are needed, vs just 1. It can be stretched a little, that's fine.

`avoid too many DOM requests    use data attributes when needed instead of parsing IDs` Decent enough change, albeit I'm not sure it has much functional benefit.
`use of JavaScript data (created a mini store system)` why? This is a big chunk of complex code, what value does it add?
`avoid re-creating the image table when not necessary` Why? It needs recreation often, there's only a small subset of times when a prior table matches the shape of a current one.
`default to a placeholder images if not created yet (404)` This part is a great change, and I would happily accept straight away if it's PR'd on its own.
`avoid creating and relying on public functions` why? What benefit does that actually convey?

**Comment by micky2be** [comment]
You can hide whitespace change for better readability in the diff view.
Quotes change was mostly for consistency. Both style were used. Single quotes being more frequent.

---
`avoid too many DOM requests, and use data attributes when needed instead of parsing IDs `
> Decent enough change, albeit I'm not sure it has much functional benefit.

Simply to avoid querying the html when not necessary (micro performance), and mostly avoid string parsing prone to bugs.
Also allow possible smart query if necessary.

`use of JavaScript data (created a mini store system)` 
> Why? This is a big chunk of complex code, what value does it add?

Like any state system. Simple manipulation and smart update. 
And mostly, easy access to current state without relying on DOM API.
It's less than 50 lines of code BTW.

`avoid re-creating the image table when not necessary`
> Why? It needs recreation often, there's only a small subset of times when a prior table matches the shape of a current one.

Improves performance and avoid UI flashing for no reason.  It happens to me more often than you might think.
I can remove some of it which would shave 15 lines in `updateTable`/`fillTable` and 12 in `createStore`.

`default to a placeholder images if not created yet (404)`
> This part is a great change, and I would happily accept straight away if it's PR'd on its own.

:+1: That would still requires some changes on how the `` is created

`avoid creating and relying on public functions`
> Why? What benefit does that actually convey?

Best practice, and avoid naming conflict with other possible code.

---

I'm sorry, I knew it would be too many changes at once, mostly due to the DOM API approach.
I spent a lot of time testing, making sure all was working as intended.
The main goal was to prep the code for future update I had in mind. Including but not limited to:
- change of of `data.js` format for easier parsing
- move HTML creation from python to JS (since we have the data). Which might help contribution too.
- option to keep SD file naming
- dynamic loading of images when ready

I can look at the current code to see if I can extract bits and bots for prior PRs.
Meanwhile, I can add comments on this PR to help you understand the logic.

**Comment by mcmonkey4eva** [comment]
`You can hide whitespace change for better readability in the diff view.` No, I can't, you're changing too much in a single commit and git diff cannot render anything usable out of it. The whitespace toggle is insufficient to compensate. This needs to be split into separate commits. 1 commit per functional change. (eg 1 commit that moves css to a separate file, and a different commit that adds any changes to the css, such that the changes made to the css are visible in git history and not hidden by the move).


`avoid creating and relying on public functions - Best practice, and avoid naming conflict with other possible code.` all applicable code is in this repo, that's really not needed. It's a huge git diff that adds complexity and reduces readability/editability for no _actual_ benefit.

`Improves performance and avoid UI flashing for no reason.` makes sense (trusting that it works and applies often)

`move HTML creation from python to JS (since we have the data). Which might help contribution too.` That would be counter-productive. Ideally if anything _more_ of the HTML would be pre-generated via python directly into the `.html` file, rather than the javascript.
Several sections of the HTML _need to be_ generated by python (eg title/description/etc) (eg so that publicly posted grid views provide accurate metadata to webcrawlers).
(Not to mention, code that runs at gen-time in the python is easier to validate than the javascript).

`change of of data.js format for easier parsing` it has a carefully planned designed, if you think there's a needed change there you should open an issue about it (which is the other half of what's wrong with this PR: where's the issues requesting any of these changes?)

`option to keep SD file naming` the primary blocker on that is just choosing how to fit the data in a performant way - ie should it be in `data.js`, or should it be in sub-files emitted alongside the images? Both options have pros and cons.

`dynamic loading of images when ready` as in, while generating? If so, that'd be a nice feature to add. Should post an issue for it.

-----

Tone gets weird on the internet sometimes, so to be clear: I appreciate the PR, and it's clear you're doing good work for good reasons.
The main thing I need right now is just to organize the git side of things such that I can clearly see what exactly it is you changed and why, independently, rather than git just showing basically total file replacements. The two things for that are just properly separating commits, and removing unnecessary mass-replacements.
The secondary thing is just, remember the difference but what's *subjectively an improvement to you* vs. what's objectively an improvement to the project. If it's an objective improvement to the code (eg the UI flashing fix, the placeholder images, etc), we can figure out how to make it work and pull it in. If it's subjective (eg reformatting, moving things around, etc), then at the end of the day I'm the one that's gonna have to deal with it long-term so only if I also subjectively prefer it (an issue is the preferred way to go about determing whether something would be good for pulling without writing it first. In cases like changing the quotes, I'd prefer an issue as it's genuinely simpler to see an issue and do the find+replace and re-testing myself than review+test a PR that does it).

File: assets/page.html (hunk 6 above)
Comment (micky2be) [reply assets/page.html:163]: Since the images have a width and height attribute now, the page is not "jumping" anymore

**Comment by mcmonkey4eva** [comment]
Closing this in favor of #55
